File: python/kickstarter/enums/error_enum.py
# Copyright (c) 2024 Jeffrey Jonathan Jennings
# Author: Jeffrey Jonathan Jennings (J3)

from enum import Enum


class error_enum(Enum):
    # Enumerators with their error codes and messages
    ERR_CODE_MISSING_OR_INVALID_FIELD = ("ERR_CODE_MISSING_OR_INVALID_FIELD", "")
    ERR_CODE_IO_EXCEPTION = ("ERR_CODE_IO_EXCEPTION", "")
    ERR_CODE_AWS_REQUESTED_SECRET_DOES_HAVE_NOT_STRING = ("ERR_CODE_AWS_REQUESTED_SECRET_DOES_HAVE_NOT_STRING", "")
    ERR_CODE_AWS_REQUESTED_SECRET_HAS_NO_VALUE = ("ERR_CODE_AWS_REQUESTED_SECRET_HAS_NO_VALUE", "")
    ERR_CODE_AWS_REQUESTED_SECRET_NOT_FOUND = ("ERR_CODE_AWS_REQUESTED_SECRET_NOT_FOUND", "")
    ERR_CODE_AWS_INVALID_REQUEST = ("ERR_CODE_AWS_INVALID_REQUEST", "")
    ERR_CODE_AWS_INVALID_PARAMETERS = ("ERR_CODE_AWS_INVALID_PARAMETERS", "")
    ERR_CODE_AWS_REQUESTED_PARAMETER_PREFIX_NOT_FOUND = ("ERR_CODE_AWS_REQUESTED_PARAMETER_PREFIX_NOT_FOUND", "")

    _BY_ERROR_CODE = {}
    _BY_ERROR_MESSAGE = {}

    def __init__(self, error_message_code, error_message):
        self.error_message_code = error_message_code
        self.error_message = error_message

    @classmethod
    def initialize_mappings(cls):
        """
        Initializes the mappings for quick lookup by error code and message.
        """
        for error in cls:
            cls._BY_ERROR_CODE[error.error_message_code] = error
            cls._BY_ERROR_MESSAGE[error.error_message] = error

    @classmethod
    def value_of_error_message_code(cls, error_message_code):
        """
        This method returns the enum value by the error_message_code.

        :param error_message_code: The error code.
        :return: The corresponding enum value.
        """
        return cls._BY_ERROR_CODE.get(error_message_code)

    @classmethod
    def value_of_error_message(cls, error_message):
        """
        This method returns the enum value by the error_message.

        :param error_message: The error message.
        :return: The corresponding enum value.
        """
        return cls._BY_ERROR_MESSAGE.get(error_message)

    def get_code(self):
        """
        :return: The error code.
        """
        return self.error_message_code

    def get_message(self):
        """
        :return: The error message.
        """
        return self.error_message


# Initialize mappings
error_enum.initialize_mappings()

# Example usage for testing purposes
if __name__ == "__main__":
    error = error_enum.value_of_error_message_code("ERR_CODE_IO_EXCEPTION")
    print(f"Error Code: {error.get_code() if error else 'Not Found'}, Message: {error.get_message() if error else 'Not Found'}")

File: java/app/src/main/java/apache_flink/kickstarter/enums/error_enum.java
/**
 * Copyright (c) 2024 Jeffrey Jonathan Jennings
 * 
 * @author Jeffrey Jonathan Jennings (J3)
 * 
 * Defined enumerators
 */
package apache_flink.kickstarter.enums;


import java.util.*;


public enum error_enum {
    // --- The constructor is invoked by the enums below
    ERR_CODE_MISSING_OR_INVALID_FIELD("ERR_CODE_MISSING_OR_INVALID_FIELD", ""),
    ERR_CODE_IO_EXCEPTION("ERR_CODE_IO_EXCEPTION", ""),
    ERR_CODE_AWS_REQUESTED_SECRET_DOES_HAVE_NOT_STRING("ERR_CODE_AWS_REQUESTED_SECRET_DOES_HAVE_NOT_STRING", ""),
    ERR_CODE_AWS_REQUESTED_SECRET_HAS_NO_VALUE("ERR_CODE_AWS_REQUESTED_SECRET_HAS_NO_VALUE", ""),
    ERR_CODE_AWS_REQUESTED_SECRET_NOT_FOUND("ERR_CODE_AWS_REQUESTED_SECRET_NOT_FOUND", ""), 
    ERR_CODE_AWS_INVALID_REQUEST("ERR_CODE_AWS_INVALID_REQUEST", ""),
    ERR_CODE_AWS_INVALID_PARAMETERS("ERR_CODE_AWS_INVALID_PARAMETERS", ""),
    ERR_CODE_AWS_REQUESTED_PARAMETER_PREFIX_NOT_FOUND("ERR_CODE_AWS_REQUESTED_PARAMETER_PREFIX_NOT_FOUND", "");

    // --- Avoid iterating the enum values by using a Map to cache them
    private static final Map<String, error_enum> _BY_ERROR_CODE = new HashMap<>();
    private static final Map<String, error_enum> _BY_ERROR_MESSAGE = new HashMap<>();

    private String errorMessageCode;
    private String errorMessage;

    static {
        for (error_enum error_enum : values()) {
            _BY_ERROR_CODE.put(error_enum.errorMessageCode, error_enum);
            _BY_ERROR_MESSAGE.put(error_enum.errorMessage, error_enum);
        }
    }

    /**
     * Default constructor method.
     *
     * @param errorMessageCode
     * @param errorMessage
     */
    private error_enum(String errorMessageCode, String errorMessage) {
        this.errorMessageCode = errorMessageCode;
        this.errorMessage = errorMessage;
    }

    /**
     * This method returns the enum value by the errorMessageCode.
     *
     * @param errorMessageCode
     * @return
     */
    public static error_enum valueOferrorMessageCode(String errorMessageCode) {
        return _BY_ERROR_CODE.get(errorMessageCode);
    }

    /**
     * This method returns the enum value by the errorMessage.
     *
     * @param errorMessage
     * @return The enum value by the errorMessage.
     */
    public static error_enum valueOfErrorMessage(String errorMessage) {
        return _BY_ERROR_MESSAGE.get(errorMessage);
    }

    /**
     * @return The error code.
     */
    public String getCode() {
        return this.errorMessageCode;
    }

     /**
     * @return The error message.
     */
    public String getMessage() {
        return this.errorMessage;
    }
}